torchdata/dataloader2/dataloader2.py

```diff
@@ -365,6 +381,28 @@ def _restore_checkpoint_beginning_of_epoch(self) -> None:
         """
         self._seed_generator = self._initial_seed_generator
 
+    def _restore_naive_checkpoint(self, num_prev_yielded_batches: Optional[int] = None) -> DataLoader2Iterator[T_co]:
```

Comment:
When do we need to specify `num_prev_yielded_batches`?

I have a noob question: Why don't we make it automatically choosing the restoring option between naive or advanced based on if `NUM_PREV_YIELDED_BATCH_KEY_NAME` in the `state_dict`?

Reply:
> When do we need to specify `num_prev_yielded_batches`?

I see it as an option to override or to restore even if `self._num_prev_yielded_batches` has not been set. We can definitely take it out if you don't find it useful (and add it later if we see a need).

> automatically choose

I can imagine there are situations where users only want to restore the randomness state:
1) the model is only saved once per epoch (so you want DataLoader2 to be in sync)
2) maybe "naive" restoration is too slow and users just want to restore the randomness state, then do something custom

Reply:
> I see it as an option to override or to restore even if `self._num_prev_yielded_batches` has not been set.

If needed, users can always skipping number of iterations after doing `_restore_checkpoint_beginning_of_epoch`, right? I would like to keep the API minimum at first until a solid use case is required.


> * the model is only saved once per epoch (so you want DataLoader2 to be in sync)

In this case, we only need to save dataloader state once per epoch as well, right? If we want to support fault-tolerant, we need to make sure model has been stored at the same time.

> * maybe "naive" restoration is too slow and users just want to restore the randomness state, then do something custom

Can we list all the expected scenarios and corresponding API calls in the summary of the PR?